src/Microsoft.AspNetCore.Proxy/ProxyAdvancedExtensions.cs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Proxy
+{
+    public static class ProxyAdvancedExtentions
+    {
+        private static readonly string[] NotForwardedWebSocketHeaders = new[] { "Connection", "Host", "Upgrade", "Sec-WebSocket-Key", "Sec-WebSocket-Version" };
+        private const int DefaultWebSocketBufferSize = 4096;
+
+        public static Uri ToWebSocketScheme(this Uri uri)
```

Comment:
Below is the code designed for advanced scenarios. I wanted to use callbacks as in #10 . But the more I was working with different scenarios the more those 4 helper methods seemed to be an optimal advanced abstraction.
```
using Microsoft.AspNetCore.Proxy;

[...]

if (context.WebSockets.IsWebSocketRequest)
{
    await context.AcceptProxyWebSocketRequest(wsDestinationUri.ToWebSocketScheme());
}
else
{
    using (var requestMessage = context.CreateProxyHttpRequest(destinationUri))
    using (var responseMessage = await context.SendProxyHttpRequest(requestMessage))
    {
        await context.ReceiveProxyHttpResponse(responseMessage);
    }
}
```